### PR TITLE
Add a simple script for running go e2e tests against existing clusters.

### LINKED ISSUE #2178: Need a simple way to verify basic networking
Twice in the last two weeks I have dealt with "bugs" which turned out to be due to the fact that the fundamental requirements of pod networking were not being met.

We need a way for a user to very quickly verify that the basic assumptions are met: pods can talk to pods, master can talk to pods, pods can talk to master, services work, etc.

### PR DESCRIPTION
**Pull request by brendandburns**
Also add an even simpler script for testing the network setup.

Closes #2178

### CODE REVIEW

**Comment by brendandburns** [comment]
@thockin @filbranden

**Comment by filbranden** [comment]
-1

This is almost exactly the same as `hack/e2e-suite/goe2e.sh`.

These two scripts are bound to diverge...

Not really sure what to suggest here, but I'm sure we need to avoid the code duplication this is generating.

For now, I've been running `e2e` directly from `_output/bin/...` manually passing it the correct parameters, it doesn't really need that many... So I'm not sure we need this one so strongly right now.

**Comment by brendandburns** [comment]
What I really want is an easy one-off test for testing the network.  I can refactor this to use the goe2e.sh script.

**Comment by brendandburns** [comment]
updated to use goe2e.sh instead.

File: cluster/test-network.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+source "${KUBE_ROOT}/cluster/kube-env.sh"
+source "${KUBE_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
+
+${KUBE_ROOT}/hack/e2e-suite/goe2e.sh -tTestNetwork
```
Comment (filbranden) [inline cluster/test-network.sh]: Is this syntax right? With POSIX flags, I'd expect `-tTestNetwork` to mean `-t -T -e -s -t -N ...`, so I think you want `-t TestNetwork` instead... Does pflag really work with this syntax?

Furthermore, `-t` is currently not working, I broke it in the Ginkgo conversion, though it's the next thing on my plate to fix... I expect to have a PR to fix it real soon now.

**Comment by satnam6502** [comment]
I run individual Go e2e tests like this:
```
$ _output/release-stage/test/kubernetes/platforms/linux/amd64/e2e --v=4 --host=104.154.39.253 --provider=gce --test="TestNetwork"
I0128 16:30:16.216866   16554 driver.go:115] Skipping test 1 TestKubernetesROService
I0128 16:30:16.217566   16554 driver.go:115] Skipping test 2 TestKubeletSendsEvent
I0128 16:30:16.217597   16554 driver.go:115] Skipping test 3 TestImportantURLs
I0128 16:30:16.217620   16554 driver.go:115] Skipping test 4 TestPodUpdate
I0128 16:30:16.217648   16554 driver.go:115] Skipping test 6 TestClusterDNS
I0128 16:30:16.217671   16554 driver.go:115] Skipping test 7 TestPodHasServiceEnvVars
I0128 16:30:16.217694   16554 driver.go:115] Skipping test 8 TestBasic
I0128 16:30:16.217717   16554 driver.go:115] Skipping test 9 TestPrivate
I0128 16:30:16.217739   16554 driver.go:115] Skipping test 10 TestLivenessHttp
I0128 16:30:16.217762   16554 driver.go:115] Skipping test 11 TestLivenessExec
I0128 16:30:16.217824   16554 driver.go:138] Tests shuffled with orderseed 0x7b3ffd3
I0128 16:30:16.217849   16554 driver.go:143] Running test 1 TestNetwork
I0128 16:30:16.526730   16554 network.go:53] Creating service with name nettest-725
I0128 16:30:18.627135   16554 network.go:114] Attempt 0/60: service/pod still starting. (error: 'endpoints "nettest-725" not found')
I0128 16:30:20.667437   16554 network.go:114] Attempt 1/60: service/pod still starting. (error: 'no endpoints available for nettest-725')
I0128 16:30:22.707647   16554 network.go:114] Attempt 2/60: service/pod still starting. (error: 'no endpoints available for nettest-725')
I0128 16:30:24.747877   16554 network.go:114] Attempt 3/60: service/pod still starting. (error: 'no endpoints available for nettest-725')
I0128 16:30:26.787735   16554 network.go:114] Attempt 4/60: service/pod still starting. (error: 'no endpoints available for nettest-725')
I0128 16:30:28.828144   16554 network.go:114] Attempt 5/60: service/pod still starting. (error: 'no endpoints available for nettest-725')
I0128 16:30:30.870449   16554 network.go:122] Attempt 6/60: test still running
I0128 16:31:02.911655   16554 network.go:114] Attempt 7/60: service/pod still starting. (error: 'request [&{Method:GET URL:https://104.154.39.253/api/v1beta1/proxy/services/nettest-725/status Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[] Body:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:104.154.39.253 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil>}] failed (503) 503 Service Unavailable: Error: 'dial tcp 10.245.2.136:8080: i/o timeout'
Trying to reach: 'http://10.245.2.136:8080/status'')
I0128 16:31:34.952602   16554 network.go:114] Attempt 8/60: service/pod still starting. (error: 'request [&{Method:GET URL:https://104.154.39.253/api/v1beta1/proxy/services/nettest-725/status Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[] Body:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:104.154.39.253 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil>}] failed (503) 503 Service Unavailable: Error: 'dial tcp 10.245.2.137:8080: i/o timeout'
Trying to reach: 'http://10.245.2.137:8080/status'')
I0128 16:32:06.994116   16554 network.go:114] Attempt 9/60: service/pod still starting. (error: 'request [&{Method:GET URL:https://104.154.39.253/api/v1beta1/proxy/services/nettest-725/status Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[] Body:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:104.154.39.253 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil>}] failed (503) 503 Service Unavailable: Error: 'dial tcp 10.245.1.25:8080: i/o timeout'
Trying to reach: 'http://10.245.1.25:8080/status'')
I0128 16:32:39.035295   16554 network.go:114] Attempt 10/60: service/pod still starting. (error: 'request [&{Method:GET URL:https://104.154.39.253/api/v1beta1/proxy/services/nettest-725/status Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[] Body:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:104.154.39.253 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil>}] failed (503) 503 Service Unavailable: Error: 'dial tcp 10.245.1.25:8080: i/o timeout'
Trying to reach: 'http://10.245.1.25:8080/status'')
I0128 16:33:11.075787   16554 network.go:114] Attempt 11/60: service/pod still starting. (error: 'request [&{Method:GET URL:https://104.154.39.253/api/v1beta1/proxy/services/nettest-725/status Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[] Body:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:104.154.39.253 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil>}] failed (503) 503 Service Unavailable: Error: 'dial tcp 10.245.2.136:8080: i/o timeout'
Trying to reach: 'http://10.245.2.136:8080/status'')
I0128 16:33:13.117968   16554 network.go:122] Attempt 12/60: test still running
I0128 16:33:15.201857   16554 network.go:127] Failed on attempt 13. Cleaning up. Details:
{
	"Hostname": "ff5461ec-a74d-11e4-b028-42010af09c51",
	"Sent": {
		"ff52a9a1-a74d-11e4-b028-42010af09c51": 2,
		"ff5461ec-a74d-11e4-b028-42010af09c51": 2,
		"ff55a4d9-a74d-11e4-b028-42010af09c51": 2,
		"ff598353-a74d-11e4-b028-42010af09c51": 2,
		"ff6980a4-a74d-11e4-b028-42010af09c51": 2
	},
	"Received": {
		"ff5017a4-a74d-11e4-b028-42010af09c51": 13,
		"ff52a9a1-a74d-11e4-b028-42010af09c51": 2,
		"ff5461ec-a74d-11e4-b028-42010af09c51": 2,
		"ff598353-a74d-11e4-b028-42010af09c51": 2,
		"ff5b937c-a74d-11e4-b028-42010af09c51": 3,
		"ff6980a4-a74d-11e4-b028-42010af09c51": 2,
		"ff704d0e-a74d-11e4-b028-42010af09c51": 1
	},
	"Errors": null,
	"Log": [
		"Unable to read endpoints for default/nettest-725: endpoints \"nettest-725\" not found; will try again.",
		"Warning: unable to contact 'http://10.245.2.133:8080': 'Post http://10.245.2.133:8080/write: dial tcp 10.245.2.133:8080: i/o timeout'",
		"Warning: unable to contact 'http://10.245.2.136:8080': 'Post http://10.245.2.136:8080/write: dial tcp 10.245.2.136:8080: i/o timeout'",
		"Warning: unable to contact 'http://10.245.2.137:8080': 'Post http://10.245.2.137:8080/write: dial tcp 10.245.2.137:8080: i/o timeout'",
		"Warning: unable to contact 'http://10.245.2.133:8080': 'Post http://10.245.2.133:8080/write: dial tcp 10.245.2.133:8080: i/o timeout'",
		"Warning: unable to contact 'http://10.245.2.136:8080': 'Post http://10.245.2.136:8080/write: dial tcp 10.245.2.136:8080: i/o timeout'"
	],
	"StillContactingPeers": true
}
I0128 16:33:15.844573   16554 driver.go:149]         test 1 failed
I0128 16:33:15.844651   16554 driver.go:43] 1..1
I0128 16:33:15.844680   16554 driver.go:48] not ok 1 - TestNetwork
F0128 16:33:15.844707   16554 driver.go:160] At least one test failed
```

**Comment by brendandburns** [comment]
Actually -tTestNetwork is exactly what you need, it turns out that -t=TestNetwork will look for '=TestNetwork'

Not sure that's the correct POSIX implementation, but that's what happens.

**Comment by satnam6502** [comment]
What should be done with this PR? Should it be closed and not merged?

**Comment by brendandburns** [comment]
No, I think it should be merged.  I re-factored the "test-network.sh" script in terms of the "goe2e.sh" script so that got rid of the duplicated code.  I think we should have a separate test-network.sh command since it is a good way to have users debug their network setup if they have problems (see #2178 for context)

**Comment by thockin** [comment]
I'm OK with this.  It *is* overlap, but I think having a shorthand way for users to test their network is useful, considering how many issues we get.

**Comment by filbranden** [comment]
I was thinking about this one and I'm actually leaning to make goe2e.sh stand on its own...

I'm really reluctant to having more shell scripts around because once we have users for them then they kind of expect those to be "stable"...

Anyways, if I do get to refactor the e2e test suite to make goe2e.sh stand on its own I'll update this one to just call that one.

Cheers,
Filipe

**Pull request merged**